docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+
+For details of specific memory management related flows, please refer to [the SGX EDMM driver API spec](https://github.com/openenclave/openenclave/pull/3755/files). The public EMM APIs defined here are most likely invoked by some intermediate runtime level components for specific usages, such as dynamic heap/stack, mmap, mprotect, higher level language JIT compiler, etc.   
+
+## User Experience ##
+
+**Allocate, Deallocate Enclave Memory**
+
+The EMM provides an API, sgx_mm_alloc, for clients to request enclave memory allocations. An enclave memory allocation represents both a reserved virtual address range and a commitment of EPC pages.  EPC pages are committed for enclaves via special SGX instructions: loaded by EADD/EEXTEND before EINIT or dynamically added using EAUG followed by EACCEPT. 
+
+The sgx_mm_alloc API allows clients to specify one of three committing modes for an allocation: 
+- EMA_RESERVE, only the virtual address range is reserved. No EPC pages will be committed in this mode.  
+- EMA_COMMIT_NOW: reserves and commits physical EPC committed upon allocation. EACCEPT will be done immediately.
+- EMA_COMMIT_ON_DEMAND: EACCEPT is done on demand, see below on committing and uncommitting.
+
+An allocation, once created, will own its address range until the deallocation API, sgx_mm_dealloc, is called upon. No two active allocations can have overlapping address ranges.
+
+**Commit, Uncommit Enclave Memory**
+
+When a page in COMMIT_ON_DEMAND allocations is accessed, a page fault occurs if the page was not yet commited.  The EMM will perform EACCEPT to commit the EPC page on page fault after OS doing EAUG. 
+
+The clients can also call the EMM commit API, sgx_mm_commit, to proactively commit specific sub-regions in a COMMIT_ON_DEMAND allocation to avoid future page fault.
+
+Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page fault, can register a custom handler for page fault at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
+
+Committed pages will stay committed (regardless how they were committed) until the clients calls the uncommit API, sgx_mm_uncommit, on them or the allocation they belong to is deallocated by sgx_mm_dealloc.    
+
+**Modify Page Attributes**
+
+The EMM clients may call sgx_mm_modify/sgx_mm_modify_ex to request permissions and/or page type changes for pages in existing allocations.
+
+
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
```

Comment:
```suggestion
The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called the Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
```

Also I cannot parse the grammar of "EMA and EMA list are considered private data structure of memory manager".   structure is singular but 'are' is plural.  Are "EMA" and "EMA list" two different things?